operator/scheduler/server_test.go

```diff
@@ -0,0 +1,290 @@
+			servers: []v1alpha1.Server{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:       "foo",
+						Namespace:  "default",
+						Generation: 1,
+					},
+				},
+			},
+			expectedProtos: []*scheduler.ServerNotify{
+				{
+					Name:             "foo",
+					ExpectedReplicas: 1,
+					KubernetesMeta: &scheduler.KubernetesMeta{
+						Namespace:  "default",
+						Generation: 1,
+					},
+				},
+			},
+		},
+		{
+			name: "deleted server",
+			servers: []v1alpha1.Server{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:              "foo",
+						Namespace:         "default",
+						Generation:        1,
+						DeletionTimestamp: &now,
+					},
+					Spec: v1alpha1.ServerSpec{
+						ScalingSpec: v1alpha1.ScalingSpec{
+							Replicas: getIntPtr(2),
+						},
+					},
+				},
+			},
+			expectedProtos: []*scheduler.ServerNotify{
+				{
+					Name:             "foo",
+					ExpectedReplicas: 0,
+					KubernetesMeta: &scheduler.KubernetesMeta{
+						Namespace:  "default",
+						Generation: 1,
+					},
+				},
+			},
+		},
+		{
+			name:           "nil servers",
+			expectedProtos: []*scheduler.ServerNotify{},
+		},
+		{
+			name: "list of servers",
+			servers: []v1alpha1.Server{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:       "foo",
+						Namespace:  "default",
+						Generation: 1,
+					},
+				},
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:       "bar",
+						Namespace:  "default",
+						Generation: 2,
+					},
+				},
+			},
+			expectedProtos: []*scheduler.ServerNotify{
+				{
+					Name:             "foo",
+					ExpectedReplicas: 1,
+					KubernetesMeta: &scheduler.KubernetesMeta{
+						Namespace:  "default",
+						Generation: 1,
+					},
+				},
+				{
+					Name:             "bar",
+					ExpectedReplicas: 1,
+					KubernetesMeta: &scheduler.KubernetesMeta{
+						Namespace:  "default",
+						Generation: 2,
+					},
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			// note that responses_experiments is nill -> scheduler state is not existing
+			grpcClient := mockSchedulerGrpcClient{
+				requests_servers: []*scheduler.ServerNotify{},
+			}
+			controller := newMockControllerClient()
+			err := controller.ServerNotify(context.Background(), &grpcClient, test.servers)
+			g.Expect(err).To(BeNil())
+
+			if len(test.servers) != 0 {
+				g.Expect(len(grpcClient.requests_servers)).To(Equal(len(test.expectedProtos)))
+				g.Expect(grpcClient.requests_servers).To(Equal(test.expectedProtos))
+			} else {
+				g.Expect(len(grpcClient.requests_servers)).To(Equal(0))
+			}
+
+		})
+	}
+}
+
+func TestSubscribeServerEvents(t *testing.T) {
+	g := NewGomegaWithT(t)
+
+	type test struct {
+		name               string
+		existing_resources []client.Object
+		results            []*scheduler.ServerStatusResponse
+		noSchedulerState   bool
+	}
+
+	// note expected state is derived in the test, maybe we should be explictl about it in the future
```

Comment:
I would say this is ok. Having the expected state derived in the test here shows the **constraint** explicitly (correspondence between number of model replicas). Having the expected state declared would show the state explicitly but would "hide" the nature of the constraint -- which will be implemented manually and could be error prone.

I thus argue that in this case it might be better to have the expected state derived within the test.